sklearn_numba_dpex/kmeans/kernels/kmeans_plusplus.py

```diff
@@ -0,0 +1,224 @@
+import math
+from functools import lru_cache
+
+import numpy as np
+import numba_dpex as dpex
+
+from ._base_kmeans_kernel_funcs import (
+    _make_initialize_window_kernel_funcs,
+    _make_accumulate_sum_of_ops_kernel_func,
+)
+from sklearn_numba_dpex.common.random import make_rand_uniform_kernel_func
+
+# NB: refer to the definition of the main lloyd function for a more comprehensive
+# inline commenting of the kernel.
+
+
+@lru_cache
+def make_init_kmeansplusplus_kernel(
+    n_samples,
+    n_features,
+    preferred_work_group_size_multiple,
+    work_group_size,
+    dtype,
+):
+
+    zero_idx = np.int64(0)
+    zero_init = dtype(0.0)
+
+    @dpex.kernel
+    # fmt: off
+    def init_kmeansplusplus(
+        X_t,                      # IN READ-ONLY   (n_features, n_samples)
+        sample_weight,            # IN READ-ONLY   (n_samples,)
+        centers_t,                # OUT            (n_features, n_clusters)
+        center_indices,           # OUT            (n_clusters,)
+        closest_dist_sq,          # OUT            (n_samples,)
+    ):
+    # fmt: on
+        sample_idx = dpex.get_global_id(zero_idx)
+        if sample_idx >= n_samples:
+            return
+
+        starting_center_id_ = center_indices[zero_idx]
+
+        sq_distance = zero_init
+        for feature_idx in range(n_features):
+            diff = X_t[feature_idx, sample_idx] - X_t[feature_idx, starting_center_id_]
+            sq_distance += diff * diff
+
+        sq_distance *= sample_weight[sample_idx]
+        closest_dist_sq[sample_idx] = sq_distance
+
+        if sample_idx > zero_idx:
+            return
+
+        for feature_idx in range(n_features):
+            centers_t[feature_idx, zero_idx] = X_t[feature_idx, starting_center_id_]
+
+    global_size = (math.ceil(n_samples / work_group_size)) * (work_group_size)
+    return init_kmeansplusplus[global_size, work_group_size]
+
+
+@lru_cache
+def make_sample_center_candidates_kernel(
+    n_samples,
+    n_local_trials,
+    preferred_work_group_size_multiple,
+    work_group_size,
+    dtype,
+):
+
+    rand_uniform_kernel_func = make_rand_uniform_kernel_func(np.dtype(dtype))
+
+    zero_idx = np.int64(0)
+    one_incr = np.int32(1)
+    one_decr = np.int32(-1)
+    zero_init = dtype(0.0)
+    max_candidate_id = np.int32(n_samples - 1)
+
+    @dpex.kernel
+    # fmt: off
+    def sample_center_candidates(
+        closest_dist_sq,          # IN             (n_features, n_samples)
+        total_potential,          # IN             (1,)
```

Comment:
Naive question: what is the meaning of "potential" here? Does this describe inertia?

Reply:
I copied vocabulary from sklearn implementation. I don't know why it's called potential.

Reply:
OK, then this looks fine to me.